File: py/immpy/renderer.py
from typing import Dict, List, Optional

from mypy_utils import unwrap
from host_tree import HostTree, HostNode
from node import (
        ShadowNode, RenderNode, RenderNodeData,
        create_shadow_node_for_render_node
)

class Renderer:
    def __init__(self, shadow_root: ShadowNode):
        self.shadow_stack: List[Optional[ShadowNode]] = [shadow_root]

        render_root = RenderNode(
                RenderNodeData(key="_root_", kind="_root_", properties={})
        )

        self.render_stack: List[Optional[RenderNode]] = [render_root]

        self.reconciliation_state: Dict[ShadowNode, Optional[RenderNode]] = {
                shadow_root: render_root
        }

    def render(self, render_node: RenderNode) -> Optional[ShadowNode]:
        shadow_parent = self.shadow_stack[-1]
        render_parent = self.render_stack[-1]

        unwrap(render_parent).append_child(render_node)

        if shadow_parent is None:
            return None
        else:
            match_index = self._match(shadow_parent, render_node)
            if match_index >= 0:
                shadow_child = shadow_parent.children()[match_index]
                self.reconciliation_state[shadow_child] = render_node
                return shadow_child

        return None

    def commit(self, host_tree: HostTree, host_node: HostNode):
        if len(self.shadow_stack) != 1 or len(self.render_stack) != 1:
            raise Exception

        shadow_root = unwrap(self.shadow_stack[-1])
        self._commit_at(shadow_root, host_node, host_tree)

    def _commit_at(self, shadow_node: ShadowNode, host_node: HostNode,
            host_tree: HostTree):
        render_node = self.reconciliation_state[shadow_node]
        if render_node is None:
            raise Exception

        shadow_node.data.properties = render_node.data.properties
        host_tree.update_node_properties(host_node, shadow_node.data.properties)

        shadow_children = shadow_node.children()
        for index, child in enumerate(shadow_children):
            match = self.reconciliation_state.get(child, None)
            if match is None:
                shadow_node.remove_child_at(index)
                host_tree.remove_child_at(host_node, index)
                host_tree.destroy_node(host_node)


        render_children = render_node.children()
        for index, child in enumerate(render_children):
            matched = self._get_match_for_render_node(child)
            if matched is None:
                new_shadow_node = create_shadow_node_for_render_node(child)

                self.reconciliation_state[new_shadow_node] = child
                shadow_node.insert_child_at(new_shadow_node, index)

                new_host_node = host_tree.create_node(child.data.kind)
                host_tree.insert_child_at(host_node, new_host_node, index)


        shadow_children = shadow_node.children()
        host_children = host_tree.get_children(host_node)
        if len(shadow_children) != len(host_children):
            raise Exception

        for index, child in enumerate(shadow_children):
            self._commit_at(child, host_children[index], host_tree)

    def _get_match_for_render_node(self, node: RenderNode) -> Optional[ShadowNode]:
        for k in self.reconciliation_state:
            if self.reconciliation_state[k] == node:
                return k

        return None

    def _match(self, shadow_parent: ShadowNode, render_node: RenderNode) -> int:
        children = shadow_parent.children()
        for index, child in enumerate(children):
            match = self.reconciliation_state.get(child, None)
            if match is None and child.data.key == render_node.data.key:
                return index

        return -1

File: go/renderer.go
package immgo

import (
	"fmt"
)

// Renderer can be thought of as the data structure that holds the state of a
// render phase. It is the public-facing API for applications to render content
// to the host.
type Renderer struct {
	shadowStack *Stack[*ShadowNode]
	renderStack *Stack[*RenderNode]

	reconciliationState map[*ShadowNode]*RenderNode
}

// NewRenderer begins a Render phase at the given shadow node.
func NewRenderer(shadowRoot *ShadowNode) *Renderer {
	shadowStack := NewStack[*ShadowNode]()
	renderStack := NewStack[*RenderNode]()
	reconciliationState := make(map[*ShadowNode]*RenderNode)

	// create a dummy render node to serve as the match for the shadow root. 
	renderRoot := NewRenderNode("_root_", "_root_", Properties{})

	shadowStack.Push(shadowRoot)
	renderStack.Push(renderRoot)

	reconciliationState[shadowRoot] = renderRoot

	return &Renderer {
		shadowStack: shadowStack,
		renderStack: renderStack,
		reconciliationState: reconciliationState,
	}
}

func (this *Renderer) Render(renderNode *RenderNode) *ShadowNode {
	shadowParent := this.shadowStack.Top()
	renderParent := this.renderStack.Top()

	// In all cases, we're going to attach this render node to the current
	// render parent.
	renderParent.AppendChild(renderNode)

	// Try to match this node against the shadow tree.
	if shadowParent == nil {
		// We're in a subtree of an unmatched node. No matching can be done,
		// just move on.
		return nil
	} else {
		// Otherwise, let's try to match this node to the current shadow parent.
		matchIndex := this.match(shadowParent, renderNode)
		if matchIndex >= 0 {
			// If we found a match, record it and return it.
			shadowChild := shadowParent.Children()[matchIndex]
			this.reconciliationState[shadowChild] = renderNode
			return shadowChild
		} 
	}

	return nil
}

func (this *Renderer) PushRenderParent(renderNode *RenderNode) error {
	shadowParent := this.shadowStack.Top()
	renderParent := this.renderStack.Top()

	// Validate that this is a child of the current render parent.
	if !renderParent.HasChild(renderNode) {
		return fmt.Errorf("Renderer.PushRenderParent: argument `renderNode` is not a child of the current render parent.")
	}

	// See if this render node had a match. If so, push it as well.
	if shadowParent != nil {
		shadowChildren := shadowParent.Children()
		pushed := false
		for _, child := range shadowChildren {
			if this.reconciliationState[child] == renderNode {
				this.shadowStack.Push(child)
				pushed = true
			}
		}

		// If this render node was not matched, we push a nil value, to signify
		// for later that we are in a subtree that is unmatched.
		if !pushed {
			this.shadowStack.Push(nil)
		}
	} else {
		this.shadowStack.Push(nil)
	}

	this.renderStack.Push(renderNode)
	return nil
}

func (this *Renderer) PopRenderParent(renderNode *RenderNode) error {
	renderParent := this.renderStack.Top()

	// Validate that this is the render parent.
	if renderNode != nil && renderParent != renderNode {
		return fmt.Errorf("Renderer.PopRenderParent: argument `renderNode` is not the current render parent.")
	}

	this.shadowStack.Pop()
	this.renderStack.Pop()
	return nil
}

// Commit applies the changes from this render phase to the given host tree. 
//
// One note about the implementation here is that it creates host tree nodes
// corresponding to hook nodes. I'd like to fix that; there shouldn't be
// anything reflected in the host tree corresponding to hooks.
func (this *Renderer) Commit(hostTree HostTree, hostNode HostNode) error {
	if this.shadowStack.Len() != 1 || this.renderStack.Len() != 1 {
		return fmt.Errorf("Renderer.Commit: Not all calls to `PushRenderParent` were matched with calls to `PopRenderParent` (shadowStack: %d, renderStack: %d", this.shadowStack.Len(), this.renderStack.Len())
	}

	shadowRoot := this.shadowStack.Top()
	return this.commitAt(shadowRoot, hostNode, hostTree)
}

func (this *Renderer) commitAt(shadowNode *ShadowNode, hostNode HostNode, hostTree HostTree) error {
	renderNode := this.reconciliationState[shadowNode]
	if renderNode == nil {
		return fmt.Errorf("Internal Library Error: no match found for shadow node within commitAt")
	}

	// update our properties
	shadowNode.data.properties = renderNode.data.properties
	if err := hostTree.UpdateNodeProperties(hostNode, shadowNode.data.properties); err != nil {
		return err
	}

	// delete unmatched children
	shadowChildren := shadowNode.Children()
	hostChildren, err := hostTree.GetChildren(hostNode)
	if err != nil {
		return err
	}
	removed := 0
	for index, child := range shadowChildren {
		_, matched := this.reconciliationState[child]

		if !matched {
			// clean up shadow node
			f, exists := shadowNode.data.properties.EventHandlers["_immgo_unmount"]
			if exists {
				f(struct{}{})
			}
			
			shadowNode.RemoveChildAt(index - removed)
			err := hostTree.RemoveChildAt(hostNode, index - removed)
			if err != nil {
				return err
			}
			hostTree.DestroyNode(hostChildren[index])
			removed += 1
		} 
	}

	// reorder children (TODO: We might want to just avoid this in `match`. This
	// is currently a bug!)

	// create new children
	renderChildren := renderNode.Children()
	for index, child := range renderChildren {
		_, matched := this.getMatchForRenderNode(child)
		if !matched {
			newShadowNode := NewShadowNodeForRenderNode(child)
			// Add this to the reconciliation state for the benefit of
			// the recursion.
			this.reconciliationState[newShadowNode] = child
			shadowNode.InsertChildAt(newShadowNode, index)

			newHostNode, err := hostTree.CreateNode(child.data.kind)
			if err != nil {
				return err
			}
			err = hostTree.InsertChildAt(hostNode, newHostNode, index)
			if err != nil {
				return err
			}
		} 
	}

	// recurse down the shadow tree.
	//
	// At this point, we expect the shadow tree and host tree to be in sync at
	// this level.
	shadowChildren = shadowNode.Children()
	hostChildren, err = hostTree.GetChildren(hostNode)
	if err != nil {
		return err
	}
	if len(shadowChildren) != len(hostChildren) {
		return fmt.Errorf("Internal Library Error; host tree and shadow tree out of sync.")
	}

	for index, child := range shadowChildren {
		this.commitAt(child, hostChildren[index], hostTree)
	}

	return nil
}


// match attemps to match `renderNode` to one of `shadowParent`'s children.
//
// A match is found if **any** of `shadowParent`'s unmatched children share the
// same key as `renderNode`. 
//
// Returns the child index of the match, or -1 if none was found.
func (this *Renderer) match(shadowParent *ShadowNode, renderNode *RenderNode) int {
	children := shadowParent.Children()
	for index, child := range children {
		_, hasMatch := this.reconciliationState[child]
		if !hasMatch && child.data.key == renderNode.data.key {
			return index
		}
	}

	return -1
}

func (this *Renderer) getMatchForRenderNode(renderNode *RenderNode) (*ShadowNode, bool) {
	for k, v := range this.reconciliationState {
		if v == renderNode {
			return k, true
		}
	}

	return nil, false
}
